opm/autodiff/StandardWell_impl.hpp

```diff
@@ -2215,4 +2220,114 @@ namespace Opm
             }
         }
     }
+
+
+
+
+    template<typename TypeTag>
+    double
+    StandardWell<TypeTag>::
+    relaxationFactorFraction(const double old_value,
+                             const double dx)
+    {
+        assert(old_value >= 0. && old_value <= 1.0);
+
+        double relaxation_factor = 1.;
+
+        // updated values without relaxation factor
+        const double possible_updated_value = old_value - dx;
+
+        // 0.95 is an experimental value remains to be optimized
+        if (possible_updated_value < 0.0) {
+            relaxation_factor = std::abs(old_value / dx) * 0.95;
+        } else if (possible_updated_value > 1.0) {
+            relaxation_factor = std::abs((1. - old_value) / dx) * 0.95;
+        }
+        // if possible_updated_value is between 0. and 1.0, then relaxation_factor
+        // remains to be one
+
+        assert(relaxation_factor >= 0. && relaxation_factor <= 1.);
+
+        return relaxation_factor;
+    }
+
+
+
+
+
+    template<typename TypeTag>
+    double
+    StandardWell<TypeTag>::
+    determineRelaxationFactorProducer(const std::vector<double>& primary_variables,
+                                      const BVectorWell& dwells)
+    {
+        // TODO: not considering solvent yet
+        // 0.95 is a experimental value, which remains to be optimized
+        double relaxation_factor = 1.0;
+
+        if (FluidSystem::phaseIsActive(FluidSystem::waterPhaseIdx)) {
+            const double relaxation_factor_w = relaxationFactorFraction(primary_variables[WFrac], dwells[0][WFrac]);
+            relaxation_factor = std::min(relaxation_factor, relaxation_factor_w);
+        }
+
+        if (FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
+            const double relaxation_factor_g = relaxationFactorFraction(primary_variables[GFrac], dwells[0][GFrac]);
+            relaxation_factor = std::min(relaxation_factor, relaxation_factor_g);
+        }
+
+        if (FluidSystem::phaseIsActive(FluidSystem::waterPhaseIdx) && FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
+            // We need to make sure the even with the relaxation_factor, the sum of F_w and F_g is below one, so there will
+            // not be negative oil fraction later
+            const double original_sum = primary_variables[WFrac] + primary_variables[GFrac];
+            const double relaxed_update = (dwells[0][WFrac] + dwells[0][GFrac]) * relaxation_factor;
+            const double possible_updated_sum = original_sum - relaxed_update;
+
+            if (possible_updated_sum > 1.0) {
+                assert(relaxed_update != 0.);
+
+                const double further_relaxation_factor = std::abs((1. - original_sum) / relaxed_update) * 0.95;
+                relaxation_factor *= further_relaxation_factor;
+            }
+        }
+
```

Comment:
Avoid double blank lines inside functions.

Reply:
It should not be. zero value rates from the `processFractions()` really mostly cause problem for the THP control mode, which has a lot of ratio values (WFR, GFR) there (non-desired zero value rates basically mean INF or NAN value for ratios). 0.95 is kind of value to make sure that we will not get INF and NAN for these rate value and also keep the convergence rate high.  If we go a little more extreme to 0.9999, we can still avoid INF and NAN, while likely we will get too big ratio values to cause VFP tables to extrapolate.  Values like 0.5, might impact convergence rate, while was not tested much.  Anyway, without testing, 0.95 is not optimal value in anyway.  And the optimal values are possibly depending on specific cases. 

Anyway, if there is some other suggestions to tangle this numerical problem (INF/NAN ratio values), I will be willing to try.

Reply:
Thanks for the explanation. Maybe you could add a sentence or two of this explanation in the comment. Then we will have a hint on where and how to tune this if this becomes an issue in the future.